src/protocol/vm/state_builder.rs

```diff
@@ -0,0 +1,461 @@
+use std::collections::{HashMap, HashSet};
+
+use alloy_primitives::Address;
+use chrono::Utc;
+use ethers::{
+    abi::{decode, ParamType},
+    prelude::U256,
+    types::H160,
+    utils::to_checksum,
+};
+use itertools::Itertools;
+use revm::{
+    precompile::{Address as rAddress, Bytes},
+    primitives::{alloy_primitives::Keccak256, AccountInfo, Bytecode, KECCAK_EMPTY},
+};
+use tracing::warn;
+
+use super::{
+    constants::ADAPTER_ADDRESS,
+    state::VMPoolState,
+    utils::{load_erc20_bytecode, ERC20Slots},
+};
+use crate::{
+    evm::{
+        engine_db_interface::EngineDatabaseInterface,
+        simulation::{SimulationEngine, SimulationParameters},
+        simulation_db::BlockHeader,
+        token,
+        tycho_db::PreCachedDB,
+        ContractCompiler,
+    },
+    protocol::{
+        errors::SimulationError,
+        vm::{
+            constants::{EXTERNAL_ACCOUNT, MAX_BALANCE},
+            engine::{create_engine, SHARED_TYCHO_DB},
+            models::Capability,
+            tycho_simulation_contract::TychoSimulationContract,
+            utils::get_code_for_contract,
+        },
+    },
+};
+
+#[derive(Debug)]
+/// `VMPoolStateBuilder` is a builder pattern implementation for creating instances of
+/// `VMPoolState`. This struct provides a flexible way to construct `VMPoolState` objects with
+/// multiple optional parameters. It handles the validation of required fields and applies default
+/// values for optional parameters where necessary.
+pub struct VMPoolStateBuilder {
+    id: String,
+    tokens: Vec<H160>,
+    block: BlockHeader,
+    balances: Option<HashMap<H160, U256>>,
+    balance_owner: Option<H160>,
+    capabilities: Option<HashSet<Capability>>,
+    involved_contracts: Option<HashSet<H160>>,
+    stateless_contracts: Option<HashMap<String, Option<Vec<u8>>>>,
+    token_storage_slots: Option<HashMap<H160, (ERC20Slots, ContractCompiler)>>,
+    manual_updates: Option<bool>,
+    trace: Option<bool>,
+    engine: Option<SimulationEngine<PreCachedDB>>,
+    adapter_contract: Option<TychoSimulationContract<PreCachedDB>>,
+    adapter_contract_path: Option<String>,
+}
+
+impl VMPoolStateBuilder {
+    pub fn new(id: String, tokens: Vec<H160>, block: BlockHeader) -> Self {
+        Self {
+            id,
+            tokens,
+            block,
+            balances: None,
+            balance_owner: None,
+            capabilities: None,
+            involved_contracts: None,
+            stateless_contracts: None,
+            token_storage_slots: None,
+            manual_updates: None,
+            trace: None,
+            engine: None,
+            adapter_contract: None,
+            adapter_contract_path: None,
+        }
+    }
+
+    pub fn balances(mut self, balances: HashMap<H160, U256>) -> Self {
+        self.balances = Some(balances);
+        self
+    }
+
+    pub fn balance_owner(mut self, balance_owner: H160) -> Self {
+        self.balance_owner = Some(balance_owner);
+        self
+    }
```

Comment:
Yes, I've thought about it for a while and my conclusion was that in a builder we pass the value `V` and not `Option<V>` because all values are `None` by default and if we call the setter it means that we want to set it to something

That's also what I've seen in other codebase (I checked revm)

I also feel that the `if let Some(...)` kills the dynamic of the builder though. But I didn't find a better solution that wasn't affecting the builder and making it a bit weird